### DISCUSSION TITLE
Whether NeMo-Guardrails supports my own locally trained llm?

### DISCUSSION THREAD

**Discussion by chenrongbinturbo** [NVIDIA-NeMo/Guardrails#124] [Q&A]
hi, can you tell me if NeMo-Guardrails supports llm with local fine tuning? If so, how should I use it?

**Comment by pranjaldoshi96** [comment]
There is an option to provide custom LLM in configuration file 
https://github.com/NVIDIA/NeMo-Guardrails/blob/main/docs/user_guide/configuration-guide.md#general-options

**Comment by trebedea** [comment]
Hi @chenrongbinturbo ,

If you are using a model based on HuggingFace `transformers`, you can use the `HuggingFacePipeline` from `langchain` without defining your custom LLM Langchain interface.  You can see here an example that load a Vicuna LLM checkpoint from the disk:

https://github.com/NVIDIA/NeMo-Guardrails/blob/cb07be6e04f345905d6cef03daf0c7ea8e11a372/examples/llm/hf_pipeline_vicuna/config.py#L103

I guess this would work for most open models. What model are you using?

**Comment by serhatgktp** [comment]
See #50. You can test whether NeMo-Guardrails works with your local LLM, but it's possible the default parsers might not be able to decode your LLMs response.

It seems that NeMo-Guardrails is currently tailored towards low-temperature responses from the LLM that follows the exact expected JSON response format.

> If you are using a model based on HuggingFace transformers, you can use the HuggingFacePipeline from langchain without defining your custom LLM Langchain interface.
> . . .
> I guess this would work for most open models.

Contrary to what @trebedea said above, I haven't been able to get NeMo-Guardrails to work with any Hugging Face LLM yet. If anyone has had success with any open-source models, I would be very grateful to know about the model(s) and configurations they used.

**Comment by drazvan** [comment]
@serhatgktp: have you tried running the examples from: https://github.com/NVIDIA/NeMo-Guardrails/tree/main/examples/llm? 
Are those working for you?

**Comment by yuanboww** [comment]
Hi, is it possible to call an API end point (where my llm is served) as the main engine for nemo guardrail?

**Reply by drazvan** [reply]
Yes, definitely. Is it a custom engine? Maybe it's already supported in LangChain, which means you can use it out of the box with NeMo Guardrails.

**Reply by yuanboww** [reply]
i don't think it's supported in LangChain already - do you have an example of how to use a random api end point for NeMo Guardrails? For example, this is how we get response from the custom LLM end point
```    
response = requests.post(
        API_URL.format(xxx),
        json={xxx},
        headers=HEADERS,
    )
```

**Reply by drazvan** [reply]
You can check out the NeMo LLM one. This is the key line: https://github.com/NVIDIA/NeMo-Guardrails/blob/develop/nemoguardrails/llm/providers/nemollm.py#L264. There's a lot of boilerplate, but it should serve as a good example.

**Comment by chozillla** [comment]
So could I just have the API call my llama.cpp server instead of openAI?

**Reply by trebedea** [reply]
Yes, @chozillla - you can use llama.cpp models instead of OpenAI.
See the discussion here, we have an open MR that will make using llama.cpp models even simpler, without needing to modify the code:
https://github.com/NVIDIA/NeMo-Guardrails/discussions/1070
